### PR TITLE
Study Cards

### PR DESCRIPTION
**Pull request by sebmarste**
first implementation of the study card feature (demo)

### CODE REVIEW

File: app/build.gradle [hunk 1]
```diff
@@ -42,6 +43,9 @@ android {
         exclude 'META-INF/maven/com.google.guava/guava/pom.xml'
     }
 
+    dataBinding {
+        enabled = true
```
Comment (kordianbruck) [inline app/build.gradle]: I didn't know about this stuff - cool! :+1:

File: app/src/main/java/de/tum/in/tumcampusapp/activities/CardsDetailActivity.java [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+import de.tum.in.tumcampusapp.api.TUMCabeClient;
+import de.tum.in.tumcampusapp.auxiliary.Const;
+import de.tum.in.tumcampusapp.auxiliary.Utils;
+import de.tum.in.tumcampusapp.databinding.ActivityCardsDetailBinding;
+import de.tum.in.tumcampusapp.exceptions.NoPrivateKey;
+import de.tum.in.tumcampusapp.models.cards.StudyCard;
+import de.tum.in.tumcampusapp.models.tumcabe.ChatMember;
+import de.tum.in.tumcampusapp.models.tumcabe.ChatVerification;
+
+public class CardsDetailActivity extends AppCompatActivity {
```
Comment (kordianbruck) [inline app/src/main/java/de/tum/in/tumcampusapp/activities/CardsDetailActivity.java]: Please extend our Generic Classes - do not directly use the AppCompat stuff.

File: app/src/main/java/de/tum/in/tumcampusapp/activities/CardsDetailActivity.java [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+                    finish();
+                }
+                return true;
+            default:
+                return super.onOptionsItemSelected(item);
+        }
+    }
+
+    private boolean save() {
+        if (card.is_valid()) {
```
Comment (kordianbruck) [inline app/src/main/java/de/tum/in/tumcampusapp/activities/CardsDetailActivity.java]: Invert these `ifs` please. e.g.:
`
if (!card.is_valid()) {
    return false;
}
`

File: app/src/main/java/de/tum/in/tumcampusapp/activities/generic/ActivityForLoadingInBackground.java [hunk 4]
```diff
@@ -38,6 +38,8 @@ public ActivityForLoadingInBackground(int layoutId) {
 		super(layoutId);
 	}
 
+	public ActivityForLoadingInBackground() { super(); }
```
Comment (kordianbruck) [inline app/src/main/java/de/tum/in/tumcampusapp/activities/generic/ActivityForLoadingInBackground.java]: Why do we need this?

File: app/src/main/java/de/tum/in/tumcampusapp/activities/generic/BaseActivity.java [hunk 5]
```diff
@@ -50,12 +50,21 @@ public BaseActivity(int layoutId) {
         mLayoutId = layoutId;
     }
 
+    public BaseActivity() { mLayoutId = null;}
```
Comment (kordianbruck) [inline app/src/main/java/de/tum/in/tumcampusapp/activities/generic/BaseActivity.java]: ehhhm ??? Please explain.

File: app/src/main/java/de/tum/in/tumcampusapp/activities/generic/BaseActivity.java [hunk 6]
```diff
@@ -34,7 +34,7 @@
     /**
      * Default layouts for user interaction
      */
-    private final int mLayoutId;
+    protected final Integer mLayoutId;
```
Comment (kordianbruck) [inline app/src/main/java/de/tum/in/tumcampusapp/activities/generic/BaseActivity.java]: Why would you need an `Integer`?

File: app/src/main/java/de/tum/in/tumcampusapp/activities/generic/ProgressActivity.java [hunk 7]
```diff
@@ -51,10 +51,18 @@ public ProgressActivity(int layoutId) {
         super(layoutId);
     }
 
+    public ProgressActivity() { super(); }
```
Comment (kordianbruck) [inline app/src/main/java/de/tum/in/tumcampusapp/activities/generic/ProgressActivity.java]: ?

File: app/src/main/java/de/tum/in/tumcampusapp/activities/generic/ProgressActivity.java [hunk 8]
```diff
@@ -51,10 +51,18 @@ public ProgressActivity(int layoutId) {
...
+
     @Override
     protected void onCreate(Bundle savedInstanceState) {
         super.onCreate(savedInstanceState);
 
+        if (this.mLayoutId != null) {
```
Comment (kordianbruck) [inline app/src/main/java/de/tum/in/tumcampusapp/activities/generic/ProgressActivity.java]: This shouldn't be allowed to be null in the first place - thats why we don't have any default constructors.

File: app/src/main/java/de/tum/in/tumcampusapp/models/cards/StudyCard.java [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+import com.google.gson.annotations.SerializedName;
+
+import de.tum.in.tumcampusapp.BR;
+import de.tum.in.tumcampusapp.models.tumcabe.ChatMember;
+
+
+public class StudyCard extends BaseObservable {
+
+    private int id;
+    //private ChatMember member;
```
Comment (kordianbruck) [inline app/src/main/java/de/tum/in/tumcampusapp/models/cards/StudyCard.java]: Commented code should not be in git.

File: app/src/main/java/de/tum/in/tumcampusapp/widgets/MVVWidget.java [hunk 10]
```diff
@@ -127,6 +127,7 @@ static void updateAppWidgets(Context context, AppWidgetManager appWidgetManager,
     static void updateAppWidget(Context context, AppWidgetManager appWidgetManager,
                                 int appWidgetId, boolean forceLoadData) {
         // Get the settings for this widget from the database
+        if (transportManager == null) transportManager = new TransportManager(context);
```
Comment (kordianbruck) [inline app/src/main/java/de/tum/in/tumcampusapp/widgets/MVVWidget.java]: Bug fix? Please put this in a separate PR.

File: build.gradle [hunk 11]
```diff
@@ -18,10 +18,11 @@ allprojects {
         options.fork = true  // Fork your compilation into a child process
         options.forkOptions.setMemoryMaximumSize("4g") // Set maximum memory to 4g
         options.compilerArgs << "-Xlint:all"
-        options.compilerArgs << "-Werror"
+        // options.compilerArgs << "-Werror"
```
Comment (kordianbruck) [inline build.gradle]: Never touch this file. @pfent will be very angry...

File: app/src/main/java/de/tum/in/tumcampusapp/activities/CardsDetailActivity.java (hunk 2 above)
Comment (sebmarste) [inline app/src/main/java/de/tum/in/tumcampusapp/activities/CardsDetailActivity.java]: done

File: app/src/main/java/de/tum/in/tumcampusapp/activities/CardsDetailActivity.java (hunk 3 above)
Comment (sebmarste) [inline app/src/main/java/de/tum/in/tumcampusapp/activities/CardsDetailActivity.java]: done

File: app/src/main/java/de/tum/in/tumcampusapp/activities/generic/ActivityForLoadingInBackground.java (hunk 4 above)
Comment (sebmarste) [inline app/src/main/java/de/tum/in/tumcampusapp/activities/generic/ActivityForLoadingInBackground.java]: refactored all this (using overwriting the setUpLayout method)

File: app/src/main/java/de/tum/in/tumcampusapp/activities/generic/BaseActivity.java (hunk 6 above)
Comment (sebmarste) [inline app/src/main/java/de/tum/in/tumcampusapp/activities/generic/BaseActivity.java]: done

File: app/src/main/java/de/tum/in/tumcampusapp/activities/generic/BaseActivity.java (hunk 5 above)
Comment (sebmarste) [inline app/src/main/java/de/tum/in/tumcampusapp/activities/generic/BaseActivity.java]: removed

File: app/src/main/java/de/tum/in/tumcampusapp/activities/generic/ProgressActivity.java (hunk 7 above)
Comment (sebmarste) [inline app/src/main/java/de/tum/in/tumcampusapp/activities/generic/ProgressActivity.java]: removed

File: app/src/main/java/de/tum/in/tumcampusapp/activities/generic/ProgressActivity.java (hunk 8 above)
Comment (sebmarste) [inline app/src/main/java/de/tum/in/tumcampusapp/activities/generic/ProgressActivity.java]: removed (this was to inflate the layout with data binding)

File: build.gradle (hunk 11 above)
Comment (sebmarste) [inline build.gradle]: the problem here is, that i get warnings (i assume caused by the data binding library) and with this flag set it will not compile with warnings... ill investigate this after the presentation

Comment (kordianbruck) [inline build.gradle]: @pfent is there not some config file to add exceptions for those libs? I remember seeing it _somewhere_ but not exactly...

Comment (pfent) [inline build.gradle]: Usually, you can always `@SuppressWarnings` of the specific thing that errors.
Please don't disable the `-Werror`, it really helps to keep developers to think about their code :heart:

File: app/src/main/java/de/tum/in/tumcampusapp/models/cards/StudyCard.java (hunk 9 above)
Comment (sebmarste) [inline app/src/main/java/de/tum/in/tumcampusapp/models/cards/StudyCard.java]: done

File: build.gradle (hunk 11 above)
Comment (sebmarste) [inline build.gradle]: @pfend, @kordianbruck: This are the actual warnings I get:
```
Warning:No processor claimed any of these annotations: java.lang.SafeVarargs,android.databinding.BindingBuildInfo,org.simpleframework.xml.ElementList,android.support.annotation.NonNull,com.google.gson.annotations.SerializedName,org.simpleframework.xml.Root,retrofit2.http.PUT,android.annotation.SuppressLint,retrofit2.http.DELETE,retrofit2.http.GET,retrofit2.http.Path,org.simpleframework.xml.Text,android.support.annotation.Nullable,retrofit2.http.POST,retrofit2.http.Body,org.simpleframework.xml.Element,android.annotation.TargetApi,android.databinding.Bindable

/Users/sebastianstein/Google Drive/Uni/android praktikum/TumCampusApp/app/build/generated/source/apt/debug/de/tum/in/tumcampusapp/databinding/ActivityCardsQuizBinding.java
Warning:(59, 37) [cast] redundant cast to String
Warning:(85, 36) [cast] redundant cast to String
Warning:(180, 112) [unchecked] unchecked cast
required: ObservableField<Boolean>
found:    Object

/Users/sebastianstein/Google Drive/Uni/android praktikum/TumCampusApp/app/build/generated/source/apt/debug/de/tum/in/tumcampusapp/databinding/ActivityCardsDetailBinding.java
Warning:(52, 32) [cast] redundant cast to String
Warning:(78, 37) [cast] redundant cast to String
Warning:(104, 36) [cast] redundant cast to String
```
The first one appears since i enabled databinding for the project.
The others are cast warnings in the Binding Classes which are generated by the databinding library.
So I cant simply fix the cast problems there or add the `@SuppressWarnings` annotation.

But I totally understand why this flag is set and i am going to investigate on this after the "hot time" of the praktikum.

Comment (kordianbruck) [inline build.gradle]: I think I remember correctly, you can exculde this with proguard: 
https://github.com/krschultz/android-proguard-snippets/tree/master/libraries
https://github.com/TCA-Team/TumCampusApp/blob/master/proguard/proguard-project-rules.pro

using the -dontwarn flag

**Comment by kordianbruck** [comment]
I think we should merge this for now and finish work on this in the coming semester.

File: app/build.gradle [hunk 12]
```diff
@@ -107,7 +107,7 @@ dependencies {
     annotationProcessor "android.arch.lifecycle:compiler:$roomVersion"
     annotationProcessor "android.arch.persistence.room:compiler:$roomVersion"
     implementation "com.google.android.gms:play-services-location:$firebaseVersion"
-    implementation 'com.android.support.constraint:constraint-layout:1.0.2'
+    implementation 'com.android.support.constraint:constraint-layout:1.0.2'openinghours
```
Comment (pfent) [inline app/build.gradle]: no, please don't

Comment (kordianbruck) [reply app/build.gradle]: upps
